stablehlo/reference/Ops.cpp

```diff
@@ -76,6 +76,97 @@ Tensor eval(CosineOp op, const Tensor &operand) {
   return result;
 }
 
+Tensor eval(DotGeneralOp op, const Tensor &lhs, const Tensor &rhs) {
+  Tensor result(op.getType());
+  ArrayRef<int64_t> lhsShape = lhs.getType().getShape();
+  ArrayRef<int64_t> lhsBatchingDims =
+      op.getDotDimensionNumbers().getLhsBatchingDimensions();
+  ArrayRef<int64_t> rhsBatchingDims =
+      op.getDotDimensionNumbers().getRhsBatchingDimensions();
+  ArrayRef<int64_t> lhsContractingDims =
+      op.getDotDimensionNumbers().getLhsContractingDimensions();
+  ArrayRef<int64_t> rhsContractingDims =
+      op.getDotDimensionNumbers().getRhsContractingDimensions();
+  std::vector<int64_t> lhsNonContractingDims;
+  for (int64_t i = 0; i < lhs.getType().getRank(); ++i)
+    if (std::find(lhsBatchingDims.begin(), lhsBatchingDims.end(), i) ==
+            lhsBatchingDims.end() &&
+        std::find(lhsContractingDims.begin(), lhsContractingDims.end(), i) ==
+            lhsContractingDims.end())
+      lhsNonContractingDims.push_back(i);
+  std::vector<int64_t> rhsNonContractingDims;
+  for (int64_t i = 0; i < rhs.getType().getRank(); ++i)
+    if (std::find(rhsBatchingDims.begin(), rhsBatchingDims.end(), i) ==
+            rhsBatchingDims.end() &&
+        std::find(rhsContractingDims.begin(), rhsContractingDims.end(), i) ==
+            rhsContractingDims.end())
+      rhsNonContractingDims.push_back(i);
+  std::vector<int64_t> contractingDimSizes;
+  for (uint64_t i = 0; i < lhsContractingDims.size(); ++i)
```

Comment:
fuse the two loops to collect contractingDimSizes and compute   totalContractingSize